text/0192-remove-constructs-compat.md

```diff
@@ -0,0 +1,667 @@
+synthesis.
+
+### Motivation
+
+The reason this is not available at the base class is because the abstraction
+did not "hold water" as the AWS CDK evolved and new CDKs emerged.  In the AWS
+CDK, we eventually ended up with a centralized synthesis logic at the
+`Stage`-level ([TODO: ref]). The main reason was that we needed to "break" the
+recursion in various domain-specific points (e.g. stages, nested stacks) which
+meant that the generic logic of "traverse the entire tree and call `synthesize`"
+did not hold. In `cdk8s`, the support for [TODO: chart dependencies] required
+that the name of the output manifest will be determined based on the topologic
+order at the app level. Here again, the generic approach failed.
+
+In leu of those failures, we decided that there is no additional value in
+actually offering a synthesis mechanism at the `constructs` level. Each
+CDK-domain implements synthesis at the "right" level. This does not mean that
+specific domains can't offer a decentralized approach (i.e. call a method called
+"synthesize" on all constructs in the tree), it just means that this is not
+provided at the base `Construct` class.
+
+### Implications on CDK code base
+
+In the AWS CDK itself, `synthesize()` was used in three locations:
+
+1. `Stack` - creates a CFN template and adds itself to the cloud assembly manifest.
+2. `AssetStaging` - stages an asset into the cloud assembly.
+3. `TreeMetadata` - creates `tree.json` with information about the construct tree.
+
+For `Stack` and `TreeMetadata`, we will convert the generic `synthesize()`
+method to `_synthesizeTemplate()` and `_synthesizeTree()` and will call them
+from the centralized synthesis function.
+
+The `AssetStaging` construct does not really need to wait until synthesis in
+order to stage the asset. In fact, all the information required already exists
+during initialization. The only missing information is the cloud assembly output
+directory, and this information is actually known during initialization (we know
+this as soon as the CDK app is created). Therefore, the solution for
+`AssetStaging` is to move the staging logic to the constructor and use
+`Stage.of(this).outdir` to find the output directory of the current stage.
+
+### Implications on end-users
+
+Participation in synthesis is an "advanced" feature of the CDK framework and se
+assume most end-users don't use this directly.
+
+If they need "last minute processing", they would likely use `prepare()` (which
+is also being [TODO: removed]) but not `synthesize()`.
+
+The use case of emitting arbitrary files into the cloud assembly directory is
+weak. The cloud assembly is a well-defined format, and is currently "closed".
+There are no extension points that tools can identify. To that end, just writing
+files to the cloud assembly output directory does not make tons of sense. Having
+said that, it is still possible to do in the same way we plan for
```

Comment:
What do you mean? Example?

Reply:
Yes. How would I write something to the assembly directory in case i need the full constructed tree? for example the dependency graph, or maybe just a serialization of a sub-tree. (Just wild ideas, I don't really have an explicit use case..)